### PR TITLE
Add support for Ledger hardware wallet

### LINKED ISSUE #13: Add support for ledger as wallet

### PR DESCRIPTION
**Pull request by jordaaash**
Closes https://github.com/solana-labs/oyster-lending/issues/13

### CODE REVIEW

**Comment by jordaaash** [comment]
Screen Shot 2021-02-06 at 2 42 48 PM

Public key viewing (with confirmation on device) and transaction signing (manually tested for deposits, borrowing, repayment, and withdrawal) have been implemented.

Account selection hasn't been implemented. A fix for https://github.com/project-serum/spl-token-wallet/issues/59 will change derived account pubkeys.

Blind signing must be enabled in the Solana Ledger app per https://github.com/project-serum/spl-token-wallet/issues/71

File: src/wallet-adapters/ledger.ts [hunk 1]
```diff
@@ -0,0 +1,92 @@
+// modified from https://github.com/project-serum/spl-token-wallet/blob/master/src/utils/walletProvider/ledger-core.js
+// @TODO: check license for ^
```
Comment (bartosz-lipinski) [inline src/wallet-adapters/ledger.ts]: you can remove this, the code was originally from the Solana repo which is Apache.

File: src/wallet-adapters/ledger_adapter.tsx [hunk 2]
```diff
@@ -0,0 +1,85 @@
+// some code from https://github.com/project-serum/spl-token-wallet/blob/master/src/utils/walletProvider/ledger.js
+// @TODO: check license for ^
```
Comment (bartosz-lipinski) [inline src/wallet-adapters/ledger_adapter.tsx]: remove

File: src/wallet-adapters/ledger_adapter.tsx [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+
+import type Transport from "@ledgerhq/hw-transport";
+import type { Transaction } from "@solana/web3.js";
+
+import EventEmitter from "eventemitter3";
+import { PublicKey } from "@solana/web3.js";
+import TransportWebUSB from "@ledgerhq/hw-transport-webusb";
```
Comment (bartosz-lipinski) [inline src/wallet-adapters/ledger_adapter.tsx]: did you check whats better webusb or webhid

File: src/wallet-adapters/ledger.ts [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+export function getSolanaDerivationPath(account: number = 0, change: number = 0) {
+  const length = 4;
+  const derivationPath = Buffer.alloc(1 + length * 4);
+
+  let offset = 0;
+  offset = derivationPath.writeUInt8(length, offset);
+  offset = derivationPath.writeUInt32BE(harden(44), offset); // Using BIP44
+  offset = derivationPath.writeUInt32BE(harden(501), offset); // Solana's BIP44 path
+  offset = derivationPath.writeUInt32BE(harden(account), offset);
+  // @FIXME: https://github.com/project-serum/spl-token-wallet/issues/59
```
Comment (bartosz-lipinski) [inline src/wallet-adapters/ledger.ts]: Great catch!

**Review by bartosz-lipinski — APPROVED** [review]
@jordansexton this looks awesome!  Thanks, a lot this will be great feature!

File: src/wallet-adapters/ledger_adapter.tsx (hunk 3 above)
Comment (jordaaash) [reply src/wallet-adapters/ledger_adapter.tsx]: Yeah, WebHID is the newer protocol but it's hidden behind an experimental flag in Chrome per https://github.com/LedgerHQ/ledgerjs#ledgerhqhw-transport-

IMO this makes WebUSB the most widely practical choice for now, but eventually supporting (or at least providing a basic reference implementation for) WebHID, WebUSB, and Bluetooth would be ideal.

Comment (bartosz-lipinski) [reply src/wallet-adapters/ledger_adapter.tsx]: 👍

**Comment by jordaaash** [comment]
Made a final refactoring pass and removed the user confirmation step on the device to grab the public key for now. This might be something to consider once account selection is implemented, but for now I went with the lower friction approach.

**Comment by jstarry** [comment]
@jordansexton do you have permission to merge?

**Pull request merged**